community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/pipes/ExpandIntoPipe.scala

```diff
@@ -69,14 +67,64 @@ case class ExpandIntoPipe(source: Pipe,
     }
   }
 
+  private def iterateRelationships(row: ExecutionContext, state: QueryState, fromNode: Node, toNode: Node) = {
+
+    //check degree and iterate from the node with smaller degree
+    val fromDegree = state.query.nodeGetDegree(fromNode.getId, dir)
```

Comment:
@jexp then `dir` would contain `BOTH`.

But I agree on taking type into account. Since it actually matters.